c++/src/kj/compat/http.c++

```diff
@@ -3489,6 +3489,466 @@ WebSocketPipe newWebSocketPipe() {
+
+inline kj::Maybe<size_t> extractBitsFromValue(ArrayPtr<const char>&& bits) {
+  stripLeadingSpace(bits);
+  // We also want to strip the whitespace from the back, since `tryParseAs()` gets upset if we have
+  // something like "10   ".
+  while (bits.size() > 0 && (bits.back() == ' ' || bits.back() == '\t')) {
+    bits = bits.slice(0, bits.size() - 1);
+  }
+  return kj::str(bits).tryParseAs<size_t>();
+}
+
+kj::Maybe<CompressionParameters> validateCompressionConfig(UnverifiedConfig&& config,
+    bool isAgreement) {
+  // Verifies that the `config` is valid depending on whether we're validating a Request (offer) or
+  // a Response (agreement). This essentially consumes the `UnverifiedConfig` and converts it into a
+  // `CompressionParameters` struct.
+  CompressionParameters result;
+
+  KJ_IF_MAYBE(serverBits, config.serverMaxWindowBits) {
+    if (serverBits->size() == 0) {
+      // This means `server_max_window_bits` was passed without a value. Since a value is required,
+      // this config is invalid.
+      return nullptr;
+    } else {
+      KJ_IF_MAYBE(bits, extractBitsFromValue(kj::mv(*serverBits))) {
+        if (*bits < 8 || 15 < *bits) {
+          // Out of range -- invalid.
+          return nullptr;
+        }
+        if (isAgreement) {
+          result.inboundMaxWindowBits = *bits;
+        } else {
+          result.outboundMaxWindowBits = *bits;
+        }
+      } else {
+        // Invalid ABNF, expected 1*DIGIT.
+        return nullptr;
+      }
+    }
+  }
+
+  KJ_IF_MAYBE(clientBits, config.clientMaxWindowBits) {
+    if (clientBits->size() == 0) {
+      if (!isAgreement) {
+        // `client_max_window_bits` does not need to have a value in an offer, let's set it to 15
+        // to get the best level of compression.
+        result.inboundMaxWindowBits = 15;
+      } else {
+        // `client_max_window_bits` must have a value in a Response.
+        return nullptr;
+      }
+    } else {
+      KJ_IF_MAYBE(bits, extractBitsFromValue(kj::mv(*clientBits))) {
+        if (*bits < 8 || 15 < *bits) {
+          // Out of range -- invalid.
+          return nullptr;
+        }
+        if (isAgreement) {
+          result.outboundMaxWindowBits = *bits;
+        } else {
+          result.inboundMaxWindowBits = *bits;
+        }
+      } else {
+        // Invalid ABNF, expected 1*DIGIT.
+        return nullptr;
+      }
+    }
+  }
+
+  if (isAgreement) {
+    result.outboundNoContextTakeover = config.clientNoContextTakeover;
+    result.inboundNoContextTakeover = config.serverNoContextTakeover;
+  } else {
+    result.inboundNoContextTakeover = config.clientNoContextTakeover;
+    result.outboundNoContextTakeover = config.serverNoContextTakeover;
+  }
+  return kj::mv(result);
+}
+
+inline bool isValidExtensionName(StringPtr extension, StringPtr permitted) {
+  // Ensures the extension name is not followed by invalid characters.
+  // Assumes that `extension` starts with `permitted` (caller's responsibility).
```

Comment:
This assumption feels unexpected and error-prone.

I guess you designed it this way because `splitPartsWithCondition` already verifies a prefix match and so comparing those characters again would be a waste. However, it's kind of difficult to verify that all the call sites have really satisfied this. `retainClientCompressorConfig()` in particular doesn't actually contain the prefix check, but I guess it's assuming the caller has already filtered offers to the right prefix. But this all feels like something where bugs could be introduced later.

An alternative would be: Change `splitPartsWithCondition` to be just `splitParts()`, always returning all parts, with both leading and trailing whitespace already stripped. Then when you're parsing the list you do something like (pseudocode):

```
for (auto& extension: splitParts(extensions, ',')) {
  auto parts = splitParts(extension, ';');
  if (part[0] != "permessage-deflate"_kj) {
    // skip unknown extension
    continue;
  }

  // ... handle extension ...
}
```